server-utils/src/suspendable_stream.rs

```diff
@@ -0,0 +1,74 @@
+use std::time::{Duration, Instant};
+use tokio::timer::Delay;
+use std::io;
+use tokio::prelude::*;
+
+/// A wrapper type for tokio::Incoming that stops accepting new connections
+/// for a specified amount of time once `ulimit` is reached
+pub struct SuspendableStream<S> {
+	stream: S,
+	delay: Duration,
+	timeout: Option<Delay>,
+}
+
+impl<S> SuspendableStream<S> {
+	/// construct a new Suspendable stream, given tokio::Incoming
+	/// and the amount of time to pause for.
+	pub fn new(stream: S) -> Self {
+		SuspendableStream {
+			stream,
+			delay: Duration::from_millis(500),
+			timeout: None,
+		}
+	}
+}
+
+impl<S, I> Stream for SuspendableStream<S>
+	where S: Stream<Item=I, Error=io::Error>
+{
+	type Item = I;
+	type Error = ();
+
+	fn poll(&mut self) -> Result<Async<Option<Self::Item>>, ()> {
+		let mut resumed = false;
+		if let Some(mut timeout) = self.timeout.take() {
+			match timeout.poll() {
+				Ok(Async::Ready(_)) => {
+					resumed = true;
+				}
+				Ok(Async::NotReady) => {
+					self.timeout = Some(timeout);
+					return Ok(Async::NotReady);
+				}
+				Err(_) => unreachable!("Polling a delay shouldn't yield any errors; qed")
+			}
+		}
+
+		loop {
+			match self.stream.poll() {
+				Ok(item) => return Ok(item),
+				Err(ref e) => if connection_error(e) {
+					warn!("Connection Error: {:?}", e);
+					continue
+				}
+				Err(err) => {
+					self.delay = if resumed && self.delay.as_secs() < 5 {
+						self.delay * 2
```

Comment:
This is only half-correct. After succesfull attempt we should actually reset the delay to the initial value (I'd say when handling `Ok(`).
Also not sure if we actually need `resumed` flag:
1. just increase the delay by `2` on every error (you can have it capped as now, but extract that to a constant)
2. In case of `Ok(item)` in the stream - reset the delay to initial value.